### ISSUE TITLE
Error: No such native module ./charset.js

### ISSUE THREAD

**Issue by just4give** [nexe/nexe#219]
I have read all the previous issue related to 'No such native module ..'  but could not resolve my problem. Hence posting it here for expert eyes.  It's a nodejs express application and I was able to create the exe successfully but when executing, getting this below error

```
node.js:901
      throw new Error(`No such native module ${id}`);
      ^

Error: No such native module ./charset.js
    at NativeModule.require (node.js:901:13)
    at s (nexe.js:1:176)
    at nexe.js:1:367
    at nexe.js:158393:16
    at Array.forEach (native)
    at Object.__dirname.759 (nexe.js:158391:18)
    at s (nexe.js:1:316)
    at nexe.js:1:367
    at Object.__dirname.3.mime-types (nexe.js:32:18)
    at s (nexe.js:1:316)
```

Here is the nexe.js code (generated) 

```
function Negotiator(request) {
  if (!(this instanceof Negotiator)) return new Negotiator(request);
  this.request = request;
}

var set = { charset: 'accept-charset',
            encoding: 'accept-encoding',
            language: 'accept-language',
            mediaType: 'accept' };


function capitalize(string){
  return string.charAt(0).toUpperCase() + string.slice(1);
}

Object.keys(set).forEach(function (k) {
  var header = set[k],
      method = require('./'+k+'.js'),
      singular = k,
      plural = k + 's';

  Negotiator.prototype[plural] = function (available) {
    return method(this.request.headers[header], available);
  };

  Negotiator.prototype[singular] = function(available) {
    var set = this[plural](available);
    if (set) return set[0];
  };

  // Keep preferred* methods for legacy compatibility
  Negotiator.prototype['preferred'+capitalize(plural)] = Negotiator.prototype[plural];
  Negotiator.prototype['preferred'+capitalize(singular)] = Negotiator.prototype[singular];
})
```

Looking at the code, I noticed, code is trying to load some modules ( charset.js for example) from relative path ./charset.
So I placed charset.js ( from node_modules/negotiator/lib ) to root directory where .exe was created. But that did not solve the problem. Don't know what I am missing. Any help would be highly appreciated.

**Comment by just4give** [comment]
anyone faced similar type of issue ? and fix or work around ?

**Comment by jaredallard** [comment]
@just4give It's a browserify issue that's well known. It'll be fixed, well, able to be worked around via nexe v2 when I finish that.

**Comment by nightwolfz** [comment]
I solved this by removing the bazillion packages I had as dependencies and going for the simplest possible version of my app.

For example, instead of using socket.io, you can use socket.js which does not depend on binaries.

**Comment by palavrov** [comment]
Got the same problem and found that it is because of old version of https://github.com/jshttp/negotiator v0.4.9 in which package is used like this
```JS
var set = { charset: 'accept-charset',
            encoding: 'accept-encoding',
            language: 'accept-language',
            mediaType: 'accept' };
// ...
Object.keys(set).forEach(function (k) {
  var header = set[k],
      method = require('./'+k+'.js'),
// ...
})
```
which obviously prevents browserify from packaging `charset.js`

Actually I'm including
`socket.io` v1.4.6 which depends from
`engine.io` v1.6.9 which depends from
`accepts` v1.1.4 which depends from
`negotiator` v0.4.9

Note that `negotiator` is already fixed in the latest version and `accepts` is also accordingly updated so the problem will be solved if `engine.io` is upgraded to latest version of `accepts` and then `socket.io` is upgraded as well.

**Comment by palavrov** [comment]
Quick workaround is to shrinkwrap `negotiator`

npm-shrinkwrap.json
```JS
{
  "name": "ttk-web",
  "version": "1.0.0",
  "dependencies": {
    "engine.io": {
      "version": "1.6.9",
      "from": "engine.io@1.6.9",
      "resolved": "https://registry.npmjs.org/engine.io/-/engine.io-1.6.9.tgz",
      "dependencies": {
        "accepts": {
          "version": "1.3.3",
          "from": "accepts@1.3.3",
          "resolved": "https://registry.npmjs.org/accepts/-/accepts-1.3.3.tgz"
        },
        "negotiator": {
          "version": "0.6.1",
          "from": "negotiator@0.6.1",
          "resolved": "https://registry.npmjs.org/negotiator/-/negotiator-0.6.1.tgz"
        }
      }
    }
  }
}
```

But seems that `socket.io` & `nexe` still doesn't match very well, got something similar to #213:
```
nexe.js:50528
var clientSource = read(require.resolve('socket.io-client/socket.io.js'), 'utf-8');
                                ^
TypeError: require.resolve is not a function
    at Object.__dirname.506../client (nexe.js:50528:33)
    at s (nexe.js:1:316)
    at nexe.js:1:367
    at Object.__dirname.404.debug (nexe.js:26285:15)
    at s (nexe.js:1:316)
    at nexe.js:1:367
    at Object.__dirname.1.body-parser (nexe.js:4:16)
    at s (nexe.js:1:316)
    at nexe.js:1:367
    at Object.__dirname.4../app.js (nexe.js:186:11)
```

Lets see if there is a easy workaround.

**Comment by jaredallard** [comment]
@palavrov You can fix require.resolve by using the require in `global`. This uses the node.js require, as opposed to the browserify "wrapper".

**Comment by palavrov** [comment]
@jaredallard I did that for `charset.js` and it didn't help. Also this approach will need to have the dependent modules in the file system which is makes using of `nexe` questionable (i.e. to left some modules outside) or I got it wrong?

Whatever, as you can see from the referenced `socket.io` PR it is their internal mess. Monkey patching the local module fixes the problem (I'm not going to serve `socket.io-client` and supporting it makes just troubles for me) but was too tired last night to prepare PR - probably later today will do it and will post here a quick shrinkwrap workaround with the patched module until PR is merged.

I.e. I was able to pack `socket.io` and everything else in my application successfully and it works very well.

**Comment by jaredallard** [comment]
@palavrov Oh don't worry, that's not a permanent fix or a wanted one. I'm hard at work on making nexe@2 not have these issues with a possible new method of packing that shouldn't have to deal with any of these issues anymore, and if that doesn't pan through; at least just being able to use the old `nexe` method which worked with other ones.

TL;DR `require.resolve` shouldn't be not implemented, and I have no idea why browserify decided to just leave it out.

**Comment by palavrov** [comment]
Actually `browserify` did something about that https://github.com/substack/node-browserify#brequirefile-opts but I'm not sure if it can be easily used. I'm packaging with `webpack` and don't have experience with `browserify` but IMHO all these issues can be fixed easily in the other modules - in most cases it is a trivial patch.

How are you going to improve this in nexe@2? With different approach like the ones used to bundle everything nw.js, electron.js i.e. packing all modules and resources in one archive and then extracting in a temp directory on every run?

**Comment by palavrov** [comment]
Here is the final monkey patch workaround for `socket.io` i.e. `feathers-socketio` in my case which adds one more layer on top that can be easily removed:

npm-shrinkwrap.json
```JS
{
  "name": "ttk-web",
  "version": "1.0.0",
  "dependencies": {
    "feathers-socketio": {
      "version": "1.4.1",
      "from": "feathers-socketio@>=1.4.0 <2.0.0",
      "resolved": "https://registry.npmjs.org/feathers-socketio/-/feathers-socketio-1.4.1.tgz",
      "dependencies": {
        "socket.io": {
          "version": "1.4.6",
          "from": "socket.io@>=1.3.7 <2.0.0",
          "resolved": "codigi/socket.io#serve-client-fix",
          "dependencies": {
            "engine.io": {
              "version": "1.6.9",
              "from": "engine.io@1.6.9",
              "resolved": "https://registry.npmjs.org/engine.io/-/engine.io-1.6.9.tgz",
              "dependencies": {
                "accepts": {
                  "version": "1.3.3",
                  "from": "accepts@1.3.3",
                  "resolved": "https://registry.npmjs.org/accepts/-/accepts-1.3.3.tgz",
                  "dependencies": {
                      "negotiator": {
                          "version": "0.6.1",
                          "from": "negotiator@0.6.1",
                          "resolved": "https://registry.npmjs.org/negotiator/-/negotiator-0.6.1.tgz"
                      }
                  }
                }
              }
            }
          }
        }
      }
    }
  }
}
```

Also `serveClient` **MUST** be disabled:

```JS
app.configure(socketio({ serveClient: false }));
```

**Comment by yasserf** [comment]
@palavrov Your solution worked great for me, thanks!

**Comment by palavrov** [comment]
@yasserf, take in mind that in `npm-shrinkwrap.json` `socket.io` is a **fork** to the current master branch and can make you crazy if you try to update the module version in `package.json`. This is valid for `engine.io`, `accepts` and `negotiator` modules as well.

I hope that related PR (socketio/engine.io#397 and socketio/socket.io#2567) will be merged soon so the final fix will be just to disable `serveClient`

**Comment by yasserf** [comment]
@palavrov My issue was actually purely with engine.io negotiator dependency version, but thanks for the heads up!

**Comment by wtonsp** [comment]
I've tried some of the workaround in this thread, but I am still seeing an issue with I try to run nexe with my application.  Any ideas on what might be causing this error?


```
nexe.js:587
  "application/iges": {
        ^

Error: No such native module ./charset.js
    at NativeModule.require (bootstrap_node.js:435:13)
    at s (nexe.js:1:176)
    at nexe.js:1:367
    at nexe.js:60472:16
    at Array.forEach (native)
    at Object.__dirname.254 (nexe.js:60470:18)
    at s (nexe.js:1:316)
    at nexe.js:1:367
    at Object.__dirname.2.mime-types (nexe.js:53:18)
    at s (nexe.js:1:316)
    at nexe.js:1:367
    at Object.__dirname.344.../transport (nexe.js:77495:15)
    at s (nexe.js:1:316)
    at nexe.js:1:367
    at Object.__dirname.343.../transport (nexe.js:77420:15)
    at s (nexe.js:1:316)
    at nexe.js:1:367
    at Object.__dirname.341../polling-jsonp (nexe.js:77306:11)
    at s (nexe.js:1:316)
    at nexe.js:1:367
    at Object.__dirname.338../socket (nexe.js:76251:18)
    at s (nexe.js:1:316)
```

**Comment by wtonsp** [comment]
Is there a way to get this working with serveClient enabled?

**Comment by palavrov** [comment]
@wtonsp, probably this is easy doable but in my case was not necessary so didn't bother for workaround.

**Comment by calebboyd** [comment]
Please try v2. If you're still having problems open a new issue

**Comment by sumitWebonerds** [comment]
I was facing same problem but with some other  module  I did reinstall with latest and it is working for me  

**Problem** 

```
Error: Cannot find module './charset.js'
1) If you want to compile the package/file into executable, please pay attention to compilation warnings and specify a literal in 'require' call. 2) If you don't want to compile the package/file into executable and want to 'require' it from filesystem (likely plugin), specify an absolute path in 'require' call using process.cwd() or process.execPath.
    at Function.Module._resolveFilename (module.js:534:15)
    at Function.Module._resolveFilename (pkg/prelude/bootstrap.js:1269:46)
    at Function.Module._load (module.js:464:25)
    at Module.require (module.js:577:17)
    at Module.require (pkg/prelude/bootstrap.js:1153:31)
    at require (internal/module.js:11:18)
    at /snapshot/html/config/configurator-file-api/node_modules/accepts/node_modules/negotiator/lib/negotiator.js:21:16
    at Array.forEach (<anonymous>)
    at Object.<anonymous> (/snapshot/html/config/configurator-file-api/node_modules/accepts/node_modules/negotiator/lib/negotiator.js:19:18)
    at Module._compile (pkg/prelude/bootstrap.js:1243:22)
    at Object.Module._extensions..js (module.js:644:10)
    at Module.load (module.js:552:32)
    at tryModuleLoad (module.js:495:12)
    at Function.Module._load (module.js:487:3)
    at Module.require (module.js:577:17)
    at Module.require (pkg/prelude/bootstrap.js:1153:31)
```


**Solution**
cd /var/html/config/configurator-file-api/node_modules/accepts/

npm uninstall negotiator --save

npm install negotiator --save


Yes it is working
